packages/ra-core/src/controller/list/filter/AutoSubmitFilterForm.tsx

```diff
@@ -0,0 +1,251 @@
+        ...rest,
+    });
+    const { handleSubmit, getValues, reset, watch, formState } = formContext;
+    const { isValid } = formState;
+
+    // Ref tracking if there are internal changes pending, i.e. changes that
+    // should not trigger a reset
+    const formChangesPending = React.useRef(false);
+
+    // Reapply filterValues when they change externally
+    useEffect(() => {
+        const newValues = getFilterFormValues(getValues(), filterValues);
+        const previousValues = getValues();
+        if (formChangesPending.current) {
+            // The effect was triggered by a form change (i.e. internal change),
+            // so we don't need to reset the form
+            formChangesPending.current = false;
+            return;
+        }
+        if (!isEqual(newValues, previousValues)) {
+            reset(newValues);
+        }
+        // The reference to the filterValues object is not updated when it changes,
+        // so we must stringify it to compare it by value.
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [JSON.stringify(filterValues), getValues, reset]);
+
+    const onSubmit = useEvent((values: any): void => {
+        // Do not call setFilters if the form is invalid
+        if (!isValid) {
+            return;
+        }
+        formChangesPending.current = true;
+        setFilters({
+            ...filterValues,
+            ...values,
+        });
+    });
+    const debouncedOnSubmit = useDebouncedEvent(onSubmit, debounce || 0);
+
+    // Submit the form on values change
+    useEffect(() => {
+        const { unsubscribe } = watch((values, { name }) => {
+            // Check that the name is present to avoid setting filters when
+            // watch was triggered by a reset
+            if (name) {
+                debouncedOnSubmit(values);
+            }
+        });
+        return () => unsubscribe();
+    }, [watch, debouncedOnSubmit]);
+
+    const sourceContext = React.useMemo<SourceContextValue>(
+        () => ({
+            getSource: (source: string) => source,
+            getLabel: (source: string) =>
+                `resources.${resource}.fields.${source}`,
+        }),
+        [resource]
+    );
+
+    return (
+        <FormProvider {...formContext}>
+            <FormGroupsProvider>
+                <SourceContextProvider value={sourceContext}>
+                    <Component onSubmit={handleSubmit(onSubmit)}>
+                        {children}
+                    </Component>
+                </SourceContextProvider>
+            </FormGroupsProvider>
+        </FormProvider>
+    );
+};
+
+const HTMLForm = (props: React.HTMLAttributes<HTMLFormElement>) => (
+    <form {...props} />
+);
+
+export interface AutoSubmitFilterFormProps
+    extends Omit<UseFormProps, 'onSubmit' | 'defaultValues'> {
+    children: ReactNode;
+    validate?: ValidateForm;
+    debounce?: number | false;
+    resource?: string;
+    component?: React.ComponentType<any>;
```

Comment:
I'd name it `formComponent` instead, as it must be a form accepting an `onSubmit` prop. Also, we probably need to narrow down the type.